### PR TITLE
Cleanup around StackRouter

### PR DESCRIPTION
**Pull request by brentvatne**
- Make StackRouter a bit easier to read, update and add some comments
- Add an explicit test around stack and navigate

# Motivation

I was experimenting with the code and improving my understanding of it. Made a few changes here in the process so hopefully it's a bit easier for the next person to understand.

# Test plan

Run the tests. This should not have any impact on functionality.

### CODE REVIEW

File: src/routers/StackRouter.js [hunk 1]
```diff
@@ -348,6 +303,60 @@ export default (routeConfigs, stackConfig = {}) => {
         }
       }
 
+      // Handle pop-to-top behavior. Make sure this happens after children have had a chance to handle the action, so that the inner stack pops to top first.
+      if (action.type === NavigationActions.POP_TO_TOP) {
+        if (state.index === 0) {
+          return {
+            ...state,
```
Comment (ericvicenti) [inline src/routers/StackRouter.js]: this looks crazy, without an explanation of why its happening

File: src/routers/StackRouter.js [hunk 2]
```diff
@@ -100,63 +155,11 @@ export default (routeConfigs, stackConfig = {}) => {
...
-          routeName: initialRouteName,
-          key: action.key || generateKey(),
-        };
-        state = {
-          key: 'StackRouterRoot',
-          isTransitioning: false,
-          index: 0,
-          routes: [route],
-        };
+        return getInitialState(action);
```
Comment (ericvicenti) [inline src/routers/StackRouter.js]: I like this. I wonder if we should make this file way more functiony, in the name of comprehension. Create functions for each action and attempt to get `getStateForAction` down to ~100 lines or something

File: src/routers/StackRouter.js [hunk 3]
```diff
@@ -348,6 +303,60 @@ export default (routeConfigs, stackConfig = {}) => {
         }
       }
 
+      // Handle pop-to-top behavior. Make sure this happens after children have had a chance to handle the action, so that the inner stack pops to top first.
+      if (action.type === NavigationActions.POP_TO_TOP) {
```
Comment (ericvicenti) [inline src/routers/StackRouter.js]: we should probably fix this to accept a key at some point soon, same with the others

File: src/routers/__tests__/StackRouter-test.js [hunk 4]
```diff
@@ -513,6 +513,39 @@ describe('StackRouter', () => {
...
+      {
+        type: NavigationActions.NAVIGATE,
+        routeName: 'Corge',
+      },
+      state
+    );
+
+    expect(state2.index).toEqual(1);
+    expect(state2.routes[1].index).toEqual(1);
+    expect(state2.routes[1].routes[1].index).toEqual(1);
```
Comment (ericvicenti) [inline src/routers/__tests__/StackRouter-test.js]: might want to verify "Corge" route name for clarity here

File: src/routers/StackRouter.js (hunk 1 above)
Comment (brentvatne) [reply src/routers/StackRouter.js]: this actually didn't change but I can add a comment here

**Pull request merged**